### PR TITLE
Optimize memcpy for the wasm backend

### LINKED ISSUE #7639: memcpy from musl is 1.4K, 3x larger than handwritten one
We should check if it's worth the extra size. The extra 1K is a big chunk of the total wasm size (around 5%) in small codebases.

### PR DESCRIPTION
**Pull request by kripken**
This makes memcpy around 500 bytes, as opposed to 1,400 which is the size of musl's. This is noticeable on small programs where this can save 3% or so.

The memcpy impl here is a conversion from the asm.js one which we benchmarked extensively in the past, with some tweaks for wasm (like using i64s).

Note that this only affects the wasm backend, as we build this into compiler-rt.

Also fix the compilation of compiler-rt, which did not include `-D__EMSCRIPTEN__` before.

Fixes #7639

### CODE REVIEW

File: system/lib/libc/musl/src/string/memcpy.c [hunk 1]
```diff
@@ -2,11 +2,76 @@
...
+#endif
+
 void *memcpy(void *restrict dest, const void *restrict src, size_t n)
 {
 	unsigned char *d = dest;
 	const unsigned char *s = src;
 
+#ifdef __EMSCRIPTEN__
+
+  unsigned char *aligned_d_end;
```
Comment (jgravelle-google) [inline system/lib/libc/musl/src/string/memcpy.c]: Apparently this file was using tabs before, and now it's mixed tabs+spaces. Probably convert to all-tabs to minimize upstream diffs? Either way it shouldn't be mixed.

Comment (kripken) [reply system/lib/libc/musl/src/string/memcpy.c]: Thanks, fixed to tabs.

**Comment by juj** [comment]
There is a benchmark_memcpy.cpp file in the tests/ directory that has a test for different memcpy sizes.

Trying out -O3 Wasm build with and without this, the MB/sec performance looks like this: http://clbri.com/dump/musl_memcpy_vs_own/results_musl_vs_own.html . No clear winner, looks like musl is somewhat faster on 64-128byte transfers, but on larger sizes, this memcpy branch seems to be more consistent.

Reading musl memcpy implementation, it optimizes for the case when the source and destination pointers are misaligned with respect to each other. The benchmark_memcpy.cpp graph above only tests when source and destination are both maximally aligned to 8 bytes. Briefly testing unaligned src and dst pointers where src and dst are out of alignment with respect to each others as well, the results seem to be very similar still.

One note is that the implemented memcpy code does 64bit loads and stores, but the operations can occur on only 4 byte aligned memory, and not 8 byte aligned memory. Not sure if this might be a performance problem or not e.g. for ARM, or if they handle the case fast. That is, there's

```c
	if ((((uintptr_t)d) & 3) == (((uintptr_t)s) & 3)) {
		// The initial unaligned < 4-byte front.
		while ((((uintptr_t)d) & 3) && d < d_end) {
			*d++ = *s++;
		}
		aligned_d_end = (unsigned char *)(((uintptr_t)d_end) & -4);
```
after which it goes to 64bit copies. That code could be 

```c
	if ((((uintptr_t)d) & 7) == (((uintptr_t)s) & 7)) {
		// The initial unaligned < 8-byte front.
		while ((((uintptr_t)d) & 7) && d < d_end) {
			*d++ = *s++;
		}
		aligned_d_end = (unsigned char *)(((uintptr_t)d_end) & -8);
```

The checks `if (aligned_d_end >= 64) {` and `if (d_end >= 4) {` look a bit dirty: they are testing if absolute pointer addresses are large enough (for underflow/wraparound in the subtraction below?), but this could be confused for copy size tests, testing pointers against absolute address values is a bit awkward. Does not look like a good way to avoid this though, so would be good to throw in a comment about the comparison being a pointer underflow check before subtraction, rather than a copy size check.

**Comment by juj** [comment]
> Does not look like a good way to avoid this though,

On another thought, perhaps we can rely on GLOBAL_BASE being >= 64, so any valid pointer will be at least that high, so those checks could be just omitted, or perhaps in debug builds asserted?

**Comment by kripken** [comment]
Good point about the alignment. I combined the 32-bit ops into 64-bit, thinking it's nice they'd be more compact, but I forgot about the alignment. I guess I should change back to 32 bit.

I'll add a comment about the overflow checks, yeah, they might not be obvious.

About those checks, we do actually need them (tests failed without them). In asm2wasm we do guarantee static memory is at least 1024, but not in the wasm backend.

**Comment by dschuff** [comment]
Is there any reason we can't put this in a different file (e.g. maybe something like emscripten_memcpy.c) rather than just ifdefing out most of musl's memcpy. If we did that we could just build this instead of musl's memcpy.c right?

Also don't forget that pretty soon we'll be able to use the bulk memory proposal's memcpy, which would likely be a good drop-in replacement for the early bail-out to JS. That is hopefully still on track to ship with atomics.

**Comment by kripken** [comment]
@dschuff Makes sense to move to a separate file, yeah. Done. Then we can use spaces instead of tabs, too ;)

Fixed the alignment issue @juj pointed out too.

**Comment by juj** [comment]
lgtm, though now the code will do a 32-bit memcpy, whereas musl version did a 64-bit memcpy, this might regress performance. I am not able to do a repeat performance test run right now, but if you want to try, there's the file tests/benchmark_memcpy.cpp that you can compile to test. It may be helpful to remove the test sizes >= 8192 bytes since those go to the big memcpy path, to make the test run quicker.

**Pull request merged**